### PR TITLE
Added boltdb implementation to get, update and delete eni attachments

### PR DESCRIPTION
**Pull request by mythri-garaga**
### Summary

Added boltdb implementation to get, update and delete eni attachments. Updated agent to use data Client instead of state manager to persist eni attachment data

These changes cover the following scenarios
  * save attachment when receive an eni attachment payload from acs;
  * update an eni attachment when the attachment’s attached status is sent;
  * delete eni attachment when a task is reaped; or attachment has expired.

### Implementation details

 - `acs/handler`: replaced statemanger saver with dataClient to save and delete eni attachments data, updated the tests accordingly
 - `data`: added botldb implementation for SaveENIAttachment, DeleteENIAttachment and GetENIAttachments
- `engine`: added a wrapper method removeENIAttachment to remove eni attachment data if attachment is expired or status is not sent; also when task is deleted
- `eventhandler`: replaced statemanger saver with dataClient to save eni attachments data
### Testing

New tests cover the changes: yes

### Licensing

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: agent/acs/handler/attach_eni_handler_common.go [hunk 1]
```diff
@@ -42,10 +44,26 @@ func (handler *ackTimeoutHandler) handle() {
...
 		handler.state.RemoveENIAttachment(handler.mac)
 	}
 }
 
+func (handler *ackTimeoutHandler) removeENIAttachment(mac string) {
+	attachmentToRemove, ok := handler.state.ENIByMac(mac)
+	if ok {
+		attachmentId, err := utils.GetENIAttachmentId(attachmentToRemove.AttachmentARN)
+		if err != nil {
+			seelog.Errorf("Failed to get attachment id for %s: %v", attachmentToRemove.AttachmentARN, err)
```
Comment (shubham2892) [inline agent/acs/handler/attach_eni_handler_common.go:57]: error message starts with lower case

File: agent/acs/handler/attach_eni_handler_common.go [hunk 2]
```diff
@@ -42,10 +44,26 @@ func (handler *ackTimeoutHandler) handle() {
...
 	if !eniAttachment.IsSent() {
 		seelog.Warnf("Timed out waiting for ENI ack; removing ENI attachment record with MAC address: %s", handler.mac)
+		handler.removeENIAttachment(handler.mac)
 		handler.state.RemoveENIAttachment(handler.mac)
 	}
 }
 
+func (handler *ackTimeoutHandler) removeENIAttachment(mac string) {
+	attachmentToRemove, ok := handler.state.ENIByMac(mac)
+	if ok {
```
Comment (shubham2892) [inline agent/acs/handler/attach_eni_handler_common.go:54]: Should it throw an error when it's not `ok`

File: agent/acs/handler/attach_eni_handler_common.go [hunk 3]
```diff
@@ -42,10 +44,26 @@ func (handler *ackTimeoutHandler) handle() {
...
+		attachmentId, err := utils.GetENIAttachmentId(attachmentToRemove.AttachmentARN)
+		if err != nil {
+			seelog.Errorf("Failed to get attachment id for %s: %v", attachmentToRemove.AttachmentARN, err)
+		} else {
+			err = handler.dataClient.DeleteENIAttachment(attachmentId)
+			if err != nil {
+				seelog.Errorf("Failed to remove data for eni attachment %s: %v", attachmentId, err)
```
Comment (shubham2892) [inline agent/acs/handler/attach_eni_handler_common.go:61]: error message starts with lower case

File: agent/acs/handler/attach_eni_handler_common.go [hunk 4]
```diff
@@ -42,10 +44,26 @@ func (handler *ackTimeoutHandler) handle() {
 	}
 	if !eniAttachment.IsSent() {
 		seelog.Warnf("Timed out waiting for ENI ack; removing ENI attachment record with MAC address: %s", handler.mac)
+		handler.removeENIAttachment(handler.mac)
```
Comment (shubham2892) [inline agent/acs/handler/attach_eni_handler_common.go:47]: Is it possible to change the name of this function to something like `removeAttachmentFromDataClient`, it's difficult to say what is the difference between `removeEniAttachment` and `RemoveENIAttachment` in the next line.

File: agent/data/eniattachment_client_test.go [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+	res, err = testClient.GetENIAttachments()
+	assert.NoError(t, err)
+	assert.Len(t, res, 2)
+
+	assert.NoError(t, testClient.DeleteENIAttachment("test-arn"))
+	assert.NoError(t, testClient.DeleteENIAttachment("test-arn2"))
+	res, err = testClient.GetENIAttachments()
+	assert.NoError(t, err)
+	assert.Len(t, res, 0)
+}
```
Comment (shubham2892) [inline agent/data/eniattachment_client_test.go:64]: Can you also add a tests for error cases

File: agent/engine/data.go [hunk 6]
```diff
@@ -68,6 +68,21 @@ func (engine *DockerTaskEngine) removeTaskData(task *apitask.Task) {
 	}
 }
 
+func (engine *DockerTaskEngine) removeENIAttachment(mac string) {
+	attachmentToRemove, ok := engine.state.ENIByMac(mac)
+	if ok {
```
Comment (shubham2892) [inline agent/engine/data.go:73]: Should it throw an error when it's not ok

File: agent/engine/docker_task_engine.go [hunk 7]
```diff
@@ -573,6 +575,7 @@ func (engine *DockerTaskEngine) deleteTask(task *apitask.Task) {
 		if taskENI.IsStandardENI() {
 			seelog.Debugf("Task engine [%s]: removing eni %s from agent state",
 				task.Arn, taskENI.ID)
+			engine.removeENIAttachment(taskENI.MacAddress)
```
Comment (shubham2892) [inline agent/engine/docker_task_engine.go:578]: rename to `removeENIAttachmentFromDataClient` ?

File: agent/acs/handler/attach_eni_handler_common.go (hunk 4 above)
Comment (mythri-garaga) [reply agent/acs/handler/attach_eni_handler_common.go]: Agreed, renamed to removeENIAttachmentFromData.

File: agent/data/eniattachment_client_test.go (hunk 5 above)
Comment (mythri-garaga) [reply agent/data/eniattachment_client_test.go:64]: Added a new test TestSaveENIAttachmentInvalidID.

File: agent/engine/data.go (hunk 6 above)
Comment (mythri-garaga) [reply agent/engine/data.go:77]: Done.

File: agent/engine/docker_task_engine.go (hunk 7 above)
Comment (mythri-garaga) [reply agent/engine/docker_task_engine.go]: Renamed to removeENIAttachmentFromData.

File: agent/acs/handler/attach_eni_handler_common.go (hunk 2 above)
Comment (mythri-garaga) [reply agent/acs/handler/attach_eni_handler_common.go:58]: Done.

File: agent/acs/handler/attach_eni_handler_common.go (hunk 1 above)
Comment (mythri-garaga) [reply agent/acs/handler/attach_eni_handler_common.go]: keeping this consistent with other logs in the file

File: agent/acs/handler/attach_eni_handler_common.go (hunk 2 above)
Comment (shubham2892) [reply agent/acs/handler/attach_eni_handler_common.go:58]: `If ok` not required here

File: agent/engine/data.go (hunk 6 above)
Comment (shubham2892) [reply agent/engine/data.go:77]: `if ok` not required here now

File: agent/acs/handler/attach_eni_handler_common.go [hunk 8]
```diff
@@ -42,10 +44,28 @@ func (handler *ackTimeoutHandler) handle() {
...
 		seelog.Warnf("Timed out waiting for ENI ack; removing ENI attachment record with MAC address: %s", handler.mac)
+		handler.removeENIAttachmentData(handler.mac)
 		handler.state.RemoveENIAttachment(handler.mac)
 	}
 }
 
+func (handler *ackTimeoutHandler) removeENIAttachmentData(mac string) {
+	attachmentToRemove, ok := handler.state.ENIByMac(mac)
+	if !ok {
+		seelog.Warnf("Unable to retrieve ENI Attachment for mac address %s: ", mac)
```
Comment (fenxiong) [inline agent/acs/handler/attach_eni_handler_common.go:55]: nit: three logs in this function can all be error level

File: agent/data/eniattachment_client.go [hunk 9]
```diff
@@ -14,20 +14,44 @@
...
+	"github.com/aws/amazon-ecs-agent/agent/utils"
+	"github.com/boltdb/bolt"
 )
 
 func (c *client) SaveENIAttachment(eni *apieni.ENIAttachment) error {
-	// TODO: implementation
-	return nil
+	id, err := utils.GetENIAttachmentId(eni.AttachmentARN)
+	if err != nil {
+		return errors.New("failed to generate database eni attachment id")
```
Comment (fenxiong) [inline agent/data/eniattachment_client.go:28]: can you add attachment arn to the error message?

**Pull request merged**